### PR TITLE
bpf: getpeername hook implementation for socket lb

### LINKED ISSUE #9974: cilium, host reachable services: implement getpeername hook
See also: https://github.com/rook/rook/issues/4752

### PR DESCRIPTION
**Pull request by borkmann**
Depends on #11649 being merged first.

See commit msgs. Related kernel changes:
 - https://lore.kernel.org/bpf/cover.1589841594.git.daniel@iogearbox.net/
 - https://lore.kernel.org/bpf/20200513075849.20868-1-daniel@iogearbox.net/

Fixes: #9974

### CODE REVIEW

File: vagrant_box_defaults.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@
 $SERVER_BOX = "cilium/ubuntu-dev"
 $SERVER_VERSION= "172"
 $NETNEXT_SERVER_BOX= "cilium/ubuntu-next"
-$NETNEXT_SERVER_VERSION= "54"
+$NETNEXT_SERVER_VERSION= "55"
```
Comment (brb) [inline vagrant_box_defaults.rb]: AFAIK we need to bump it from a separate PR which branch has a special name (to populate VM cache) or to change this PR branch name.

File: bpf/include/bpf/builtins.h [hunk 2]
```diff
@@ -240,13 +241,86 @@ static __always_inline __nobuiltin("memcpy") void memcpy(void *d, const void *s,
...
+
+	default:
+		__throw_build_bug();
+	}
+
+	return r;
+#else
+	return __bpf_memcmp_builtin(x, y, len);
 #endif
+}
```
Comment (pchaigno) [inline bpf/include/bpf/builtins.h:299]: This builtin is missing a corresponding test case.

**Review by pchaigno — CHANGES_REQUESTED** [review]
Except for Martynas's comment and one missing test case, LGTM!

File: bpf/include/bpf/builtins.h (hunk 2 above)
Comment (borkmann) [reply bpf/include/bpf/builtins.h:299]: Agree, will add. Thanks for the review!

**Comment by pchaigno** [comment]
Does commit `bpf: properly reverse xlate svc:port tuple from wildcarded lookups` require a backport?

**Review by pchaigno — APPROVED** [review]
LGTM!

**Comment by borkmann** [comment]
> Does commit `bpf: properly reverse xlate svc:port tuple from wildcarded lookups` require a backport?

Yeah, I'll take care of it for 1.7.

**Comment by borkmann** [comment]
test-docs-please

**Comment by pchaigno** [comment]
> test-docs-please (green)

@borkmann I don't think that's required anymore since the doc tests run in GitHub Action.

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
retest-4.19

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
restest-4.9

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
retest-net-next

**Comment by borkmann** [comment]
retest-runtime

**Comment by borkmann** [comment]
retest-net-next

**Comment by borkmann** [comment]
vbox error:
```
10:59:40  Progress state: VBOX_E_OBJECT_IN_USE
10:59:40  VBoxManage: error: Machine delete failed
10:59:40  VBoxManage: error: Cannot close medium '/root/VirtualBox VMs/ubuntu-18.04.4-amd64-55_1590229760245_39345/ubuntu-18.04.4-amd64-55-disk001.vmdk' because it has 2 child media
10:59:40  VBoxManage: error: Details: code VBOX_E_OBJECT_IN_USE (0x80bb000c), component MediumWrap, interface IMedium
10:59:40  VBoxManage: error: Context: "RTEXITCODE handleUnregisterVM(HandlerArg*)" at line 162 of file VBoxManageMisc.cpp
10:59:40  VBoxManage: error: Machine 'dc5268d5-da68-4fc4-9b13-921a35692493' is not currently running
10:59:40  0%...10%...20%...30%...40%...50%...60%...70%...80%...90%...100%
10:59:40  VBoxManage: error: Machine '3c6abf7b-0259-4188-973c-e78ed91e0dac' is not currently running
10:59:42  0%...10%...20%...30%...40%...50%...60%...70%...80%...90%...100%
```

**Comment by borkmann** [comment]
retest-net-next

**Comment by borkmann** [comment]
retest-net-next

**Comment by borkmann** [comment]
retest-net-next

**Pull request merged**